api/paidAction/index.js

```diff
@@ -107,283 +202,286 @@ export default async function performPaidAction (actionType, args, context) {
-    await tx.invoice.update({
-      where: {
-        id: failedInvoice.id,
-        actionState: 'FAILED'
-      },
-      data: {
-        actionState: 'RETRYING'
-      }
-    })
-
-    // create a new invoice
-    const invoice = await createDbInvoice(actionType, args, context, invoiceArgs)
-
-    return {
-      result: await action.retry({ invoiceId: failedInvoice.id, newInvoiceId: invoice.id }, context),
-      invoice,
-      paymentMethod: 'OPTIMISTIC'
+    const subContext = { ...context, tx }
+    const failedInvoice = await models.invoice.findUnique({ where: { id: invoiceId, userId: me?.id ?? USER_ID.anon } })
+    if (!failedInvoice) {
+      throw new Error('invoice not found')
     }
-  }, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
-}
-
-const INVOICE_EXPIRE_SECS = 600
-const MAX_PENDING_PAID_ACTIONS_PER_USER = 100
 
-export async function assertBelowMaxPendingInvoices (context) {
-  const { models, me } = context
-  const pendingInvoices = await models.invoice.count({
-    where: {
-      userId: me?.id ?? USER_ID.anon,
-      actionState: {
-        notIn: PAID_ACTION_TERMINAL_STATES
-      }
+    if (
+      failedInvoice.actionState !== 'FAILED' &&
+      failedInvoice.actionState !== 'PENDING' &&
+      failedInvoice.actionState !== 'PENDING_HELD'
+    ) {
+      throw new Error(`actions is not in a retriable state: ${failedInvoice.actionState}`)
     }
-  })
 
-  if (pendingInvoices >= MAX_PENDING_PAID_ACTIONS_PER_USER) {
-    throw new Error('You have too many pending paid actions, cancel some or wait for them to expire')
-  }
-}
+    // cancel the previous invoice
+    await finalizeHodlInvoice({ data: { hash: failedInvoice.hash }, lnd, models, boss })
+    const actionType = failedInvoice.actionType
 
-export class NonInvoiceablePeerError extends Error {
-  constructor () {
-    super('non invoiceable peer')
-    this.name = 'NonInvoiceablePeerError'
-  }
-}
+    const paidAction = paidActions[actionType]
+    if (!paidAction) {
+      throw new Error(`invalid action type ${actionType}`)
+    }
 
-export async function createWrappedInvoice (actionType, args, context) {
-  // if the action has an invoiceable peer, we'll create a peer invoice
-  // wrap it, and return the wrapped invoice
-  const { cost, models, lnd, me } = context
-  const userId = await paidActions[actionType]?.invoiceablePeer?.(args, context)
-  if (!userId) {
-    throw new NonInvoiceablePeerError()
-  }
+    if (!failedInvoice) {
+      throw new Error(`missing invoice ${actionType}`)
+    }
 
-  await assertBelowMaxPendingInvoices(context)
+    const { msatsRequested, actionId, actionArgs } = failedInvoice
+    subContext.cost = msatsRequested
+    subContext.actionId = actionId
+    subContext.retryForInvoice = failedInvoice
+    subContext.forceInternal = forceInternal
+    subContext.actionAttempt = failedInvoice.actionAttempt + 1 // next attempt
```

Comment:
We track attempts like this to cycle through the receiving wallets.
Maybe a receiving wallet works fine for us, but the sender cannot pay the invoice (bad channels or whatever), so we just use the next receiving wallet when the sender retries the payment. This is where actionAttempt is used, to progress to the next wallet.